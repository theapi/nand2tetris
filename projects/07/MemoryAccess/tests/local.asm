// init SP pointing to 256
@256
D=A
@SP
M=D

// init LCL pointing to 300
@300
D=A
@LCL
M=D

// init ARG pointing to 400
@400
D=A
@ARG
M=D

// init THIS pointing to 3000
@3000
D=A
@THIS
M=D

// init THAT pointing to 3010
@3010
D=A
@THAT
M=D

// constant
@10
D=A   // Store the numeric value in D

// PUSH
@SP
A=M   // set the address to where the SP is pointing
M=D   // store the value in the stack
@SP
M=M+1 // Advance the stack pointer

// local
// POP
@SP
M=M-1  // decrement (pop) the stack pointer
A=M    // set the address to where the SP is pointing

D=M   // store popped value in D
@LCL   // set address to LCL
A=M   // use the value stored in LCL as the next address
M=D   // store the value at the address

// constant
@42
D=A   // Store the numeric value in D

// PUSH
@SP
A=M   // set the address to where the SP is pointing
M=D   // store the value in the stack
@SP
M=M+1 // Advance the stack pointer

// local
@2
D=A    // Store the index value in D
@LCL   // set address to LCL
D=D+M  // store the address of LCL + index in D
@R13  // temp store the address
M=D   // store the address in R13
// POP
@SP
M=M-1  // decrement (pop) the stack pointer
A=M    // set the address to where the SP is pointing

D=M   // store the value in D
@R13  // R13 address
A=M   // use the value stored in R13 as the next address
M=D   // store the value at the address

