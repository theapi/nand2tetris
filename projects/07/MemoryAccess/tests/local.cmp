|RAM[300]|RAM[301]|RAM[302]|RAM[256]|
|     10 |     21 |     42 |     21 |
