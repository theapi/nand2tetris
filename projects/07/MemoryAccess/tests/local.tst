
load local.asm,
output-file local.out,
compare-to local.cmp,
output-list RAM[300]%D1.6.1 RAM[302]%D1.6.1;


repeat 75 {
  ticktock;
}

output;
