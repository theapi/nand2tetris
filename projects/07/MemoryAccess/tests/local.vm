
push constant 21
pop local 1
push constant 10
pop local 0
push constant 42
pop local 2
push local 1
