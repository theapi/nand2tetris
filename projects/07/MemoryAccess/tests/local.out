|RAM[401]|RAM[402]|
